api/routes/users/put.test.js

```diff
@@ -0,0 +1,411 @@
+
+      roleTests.test(
+        'sets the role on the database model if valid',
+        async test => {
+          const role1 = model();
+          role1.get.withArgs('name').returns('good');
+          const role2 = model();
+          role2.get.withArgs('name').returns('better');
+          RoleModel.fetchAll.resolves([role1, role2]);
+
+          await handler(
+            {
+              body: { role: 'good' },
+              user: { id: '2' },
+              params: { id: '1' }
+            },
+            res
+          );
+
+          test.ok(
+            user.set.calledWith('auth_role', 'good'),
+            'user model is updated with new role'
+          );
+        }
+      );
+
+      roleTests.test(
+        'clears the role on the database model if an empty string is provided',
+        async test => {
+          await handler(
+            {
+              body: { role: '' },
+              user: { id: '2' },
+              params: { id: '1' }
+            },
+            res
+          );
+
+          test.ok(
+            user.set.calledWith('auth_role', null),
+            'user model is updated with new role'
+          );
+        }
+      );
+    });
+
+    tests.test('validates the model before saving', async validateTests => {
+      const user = model();
+
+      validateTests.beforeEach(async () => {
+        UserModel.fetch.resolves(user);
+      });
+
+      validateTests.test(
+        'sends an error if the model fails to validate',
+        async test => {
+          const err = new Error('error message');
+          user.validate.throws(err);
+
+          await handler({ body: {}, user: { id: 1 }, params: { id: 2 } }, res);
+
+          test.ok(res.status.calledWith(400), 'HTTP status set to 400');
+          test.ok(
+            res.send.calledWith('update-user-error message'),
+            'error message is sent'
+          );
+          test.ok(
+            res.send.calledAfter(res.status),
+            'error message is sent after setting status'
+          );
+          test.ok(res.end.called, 'response is terminated');
+          test.ok(
+            res.end.calledAfter(res.send),
+            'response is terminated after error message is sent'
+          );
+        }
+      );
+
+      validateTests.test('saves if validation succeeds', async test => {
+        user.validate.resolves();
+
+        await handler({ body: {}, user: { id: 1 }, params: { id: 2 } }, res);
+
+        test.ok(user.save.calledOnce, 'user model is saved');
+        test.ok(
+          user.save.calledAfter(user.validate),
+          'model is saved after validation'
+        );
+        test.ok(res.status.calledWith(204), 'HTTP status set to 200');
+        test.ok(res.send.notCalled, 'body not sent');
+        test.ok(res.end.calledOnce, 'response is terminated');
+        test.ok(
+          res.end.calledAfter(res.status),
+          'response is terminated after status is set'
+        );
+      });
+    });
+  });
+});
```

Comment:
I fear I've gone down a bad path with these unit tests, from the beginning. They are so complex. I think it implies the code under test is also too complex, or has too many dependencies.  Mocking all the database stuff is difficult, but necessary since we don't want to hit an actual database.  I don't know...  But writing unit tests takes longer than anything else, and that feels like a smell.  😕